zcash_client_backend/src/fees.rs

```diff
@@ -154,15 +206,25 @@ pub enum ChangeError<E, NoteRefT> {
         /// including the required fees.
         required: NonNegativeAmount,
     },
-    /// Some of the inputs provided to the transaction were determined to currently have no
-    /// economic value (i.e. their inclusion in a transaction causes fees to rise in an amount
-    /// greater than their value.)
+    /// Some of the inputs provided to the transaction have value less than the
+    /// marginal fee, and could not be determined to have any economic value in
+    /// the context of this input selection.
+    ///
+    /// This determination is potentially conservative in the sense that inputs
+    /// with value less than or equal to the marginal fee might be excluded, even
+    /// though in practice they would not cause the fee to increase. Inputs with
+    /// value greater than the marginal fee will never be excluded.
```

Comment:
"marginal fee" is specific to ZIP 317; I realize that we don't really support anything other than ZIP 317, but the concept of "marginal fee" may become less well defined in the presence of e.g. a fee rule change related to decoupled memos.

Reply:
It's well defined for all of the fee rules we currently support (0 for a fixed fee). If we add another fee rule, we can fix it then, but there's no hope of anticipating how this will need to change for arbitrary future fee rules. I already think we over-generalized the fee handling and should remove the options that don't actually work for transactions on the current network.

To be honest, I see the continued presence of this all of this complexity and technical debt as a holdover from when ZIP 317 was being fought against by former ECC management, and I'm quite frustrated by it. It should be long gone.